rust/tw_keypair/src/ed25519/keypair.rs

```diff
@@ -0,0 +1,67 @@
+// Copyright © 2017-2023 Trust Wallet.
+//
+// This file is part of Trust. The full Trust copyright notice, including
+// terms governing use, modification, and redistribution, is contained in the
+// file LICENSE at the root of the source code distribution tree.
+
+use crate::ed25519::{private::PrivateKey, public::PublicKey, signature::Signature, Hasher512};
+use crate::traits::{KeyPairTrait, SigningKeyTrait, VerifyingKeyTrait};
+use crate::Error;
+use tw_encoding::hex;
+
+/// Represents a pair of `ed25519` private and public keys.
+#[derive(Debug)]
+pub struct KeyPair<H: Hasher512> {
+    private: PrivateKey<H>,
+    public: PublicKey<H>,
+}
+
+impl<H: Hasher512> KeyPairTrait for KeyPair<H> {
+    type Private = PrivateKey<H>;
+    type Public = PublicKey<H>;
+
+    fn public(&self) -> &Self::Public {
+        &self.public
+    }
+
+    fn private(&self) -> &Self::Private {
+        &self.private
+    }
+}
+
+impl<H: Hasher512> SigningKeyTrait for KeyPair<H> {
+    type SigningMessage = Vec<u8>;
+    type Signature = Signature;
+
+    fn sign(&self, message: Self::SigningMessage) -> Result<Self::Signature, Error> {
```

Comment:
Pass `message` as a reference?

Reply:
This will require to make the `SigningKey::SigningMessage` associated type GAT by adding a lifetime parameter.
```rust
trait SigningKey {
  type SigningMessage<'a> where Self: 'a;
}
```
That makes the code incredibly complex.
Also passing `message: &Self::SigningMessage` won't have much impact as it turns into `message: &Vec<u8>`, so additional allocation is needed on the caller side.